### PR TITLE
API v2

### LINKED ISSUE #7: Port to Haskell?
This allows us to use the Reference implementation AST, to make sure we never generated invalid Dhall.

### LINKED ISSUE #46: [RFC]: Resolve defaults recursively
For example, for a deployment:

instead of filling each default with the obvious "empty type" of a field, recursively resolve the default values.

so instead of this:
```dhall
\(_params : {metadata : (../types/io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta.dhall)}) ->
{ apiVersion = ("apps/v1" : Text)
, kind = ("Deployment" : Text)
, metadata = _params.metadata
, spec = ([] : Optional (../types/io.k8s.api.apps.v1.DeploymentSpec.dhall))
, status = ([] : Optional (../types/io.k8s.api.apps.v1.DeploymentStatus.dhall))
} : ../types/io.k8s.api.apps.v1.Deployment.dhall
```

this:
```
{ apiVersion = ("apps/v1" : Text)
, kind = ("Deployment" : Text)
, metadata =  ../default/io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta.dhall
, spec =  ../default/io.k8s.api.apps.v1.DeploymentSpec.dhall
, status = ../default/io.k8s.api.apps.v1.DeploymentStatus.dhall
} : ../types/io.k8s.api.apps.v1.Deployment.dhall
```

This makes a "merge" based API where we compose layers of configurations with `//` way more natural. Similar to https://kustomize.io

### LINKED ISSUE #47: JSONSchemaProps.dhall references itself (Cyclic import)
When trying to create a custom resource definition, dhall errors with

```
↳ ./../dhall-kubernetes/default/io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.CustomResourceDefinition.dhall
  ↳ ./../dhall-kubernetes/types/io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.CustomResourceDefinitionSpec.dhall
    ↳ ./../dhall-kubernetes/types/io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.CustomResourceValidation.dhall
      ↳ ./../dhall-kubernetes/types/io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.JSONSchemaProps.dhall

Cyclic import: ./io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.JSONSchemaProps.dhall
```

The reason is that `JSONSchemaProps` [references itself](https://github.com/dhall-lang/dhall-kubernetes/blob/bc09fbe1b48d2ec902e0fd7a3883ae51a59e6222/types/io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.JSONSchemaProps.dhall#L5) (at least I think that's the issue)

### LINKED ISSUE #49: Transition to new Optional syntax
@f-f I'm sure this is already on your radar, but with the [deprecation of the old Optional syntax](https://github.com/dhall-lang/dhall-lang/wiki/Migration%3A-Deprecation-of-old-Optional-literal-syntax) the generation script will need to be updated do that `dhall-kubernetes` will continue to parse under the new standard.

### LINKED ISSUE #54: Generate top-level union type for all apis
Basically, make this auto-generated instead of written by hand: https://github.com/dhall-lang/dhall-kubernetes/issues/48#issuecomment-473676761

### PR DESCRIPTION
**Pull request by f-f**
I'm sorry if this is a huge diff, but we're basically swapping the contents of the repo 😄 

This PR ports the conversion script from Python to Haskell (fix #7).
The main reasons for this are that:
- the Python script was really hard to maintain for reasons like "converting from Swagger to Dhall is interleaved with string formatting"
- in Haskell we can use the `dhall` library to generate always syntactically correct Dhall AST. It's also much easier to keep an eye on correctness, because types and pattern matching. It also forces us to deal with things like cyclic imports from the get go

Things happening here:
- move defaults from `default` to `defaults` folder, as it is in `dhall-nethack`
- add `types.dhall`, `defaults.dhall`, so that one can now easily "pin a version" by just importing these two records at a specific commit/tag. They also make it really easy to access objects, e.g. `let types = https://raw.githubusercontent... sha256:... in types.Deployment`
- also add `typesUnion.dhall` (fix #54), so one is able to send to Kubernetes different objects in the same stream
- defaults are resolved recursively (fix #46): if there's an import of a "nullable" record, then it's not marked as Optional, making merging objects much easier
- switch to the new Optional syntax (fix #49)
- default objects are not lambdas anymore, and instead they just don't include the required fields (that is, the ones that are not nullable records), as suggested in https://github.com/dhall-lang/dhall-lang/issues/382#issuecomment-467065733
- for objects that are simple types we used to generate a simple lambda `\(a : Text) -> a` as a default, now we just don't generate a default (e.g. see `io.k8s.apimachinery.pkg.apis.meta.v1.Time`)
- autoformat all generated Dhall code

Things left to do before we can merge:
- [x] Thoroughly comment the code
- [x] Split code in different modules, as this is a huge module that mixes "business logic" with "json conversion"
- [x] Remove `package.yaml` and add version bounds
- [x] Remove HTTP fetching code, and take arguments with the location of the file instead
- [x] Think about how we are going to build this
- [x] Remove the `convert.py` script
- [x] Remove the `api` folder
- [x] Port the examples to the new API
- [x] Adjust the readme to remove the distinction between the "raw API" and the "nice API", now it's all "new API"
- [x] Add to the readme a FAQ section, as we need to document the pattern for #54
- [x] Reintroduce the `JSONSchemaProps` object, fixing the cyclic imports (fix #47)
- [x] Release the current version as 1.0

### CODE REVIEW

**Comment by f-f** [comment]
@arianvp I'll now take care of the first two items in the TODO list (adding comments and splitting code in modules), but otherwise all the items are up for grabs and any help would be amazing 😄 

About the build, we have two possibilities here:
1. we build the Haskell project here, run it, and see if it generates the same files as the ones that are versioned. If not we error out (a bit like we are doing for the readme and the examples)
2. or we move the Haskell project into the megarepo (as proposed by Gabriel in https://github.com/dhall-lang/dhall-kubernetes/issues/7#issuecomment-440906156), and build it there. I'd like to keep the generated files here, so this would mean separating the "lifetimes" of the generator script and the generated files

Option 2 is tempting so we keep up with all the changes in dhall-haskell, but on the other hand it would be really useful to build and generate on CI runs (to avoid breakage), so I think I'm leaning towards option 1

File: dhall-kubernetes.cabal
Comment (arianvp) [inline dhall-kubernetes.cabal]: adding version bounds here will make updating to new nixpkgs /stackage snapshots less of a headache as compiler errors turn into version bound errors

File: package.yaml
Comment (arianvp) [inline package.yaml]: I'd prefer if we just use  the cabal file and not use `package.yaml`.     Seems kind of weird to use `yaml` in a repo meant to defeat the usage of yaml :P 

If you really can't stand `.cabal` file format we can perhaps have  a look at https://github.com/dhall-lang/dhall-to-cabal

But I guess this is just here because the default stack template uses this

File: stack.yaml
Comment (arianvp) [inline stack.yaml]: It's probably worth using Nix to build the haskell package, but lets worry about that after we have an implementation working and need to worry about CI and such.   I can do my nix magic once we're done with impl.

File: src/Main.hs
Comment (arianvp) [inline src/Main.hs]: We should just get the swagger spec from  `getArgs :: IO [String]`  instead of doing a network request.

File: src/Main.hs
Comment (arianvp) [inline src/Main.hs]: I wonder if we could  poke upstream about this to improve this situation? There must be others running into this issue as well. Worth a shot I guess

**Review by arianvp — COMMENTED** [review]
Just some remarks. not much so far.

**Comment by arianvp** [comment]
Question @f-f and @ari-becker 

What format do we want `defaults.dhall` and `types.dhall` to have?
The current approach seems 'wrong' to me (https://github.com/dhall-lang/dhall-kubernetes/blob/e0a654ad3836cdaa1f063589734b342f2031a57f/types.dhall)
as k8s can have same object names in different namespaces, so we should definitely encode
the namespace in `types.dhall` and `defaults.dhall`.       (i.e.   there is `v1beta1/Deployment` and `v1/Deployment`) which are actually different

Something that resembles https://github.com/coralogix/dhall-prometheus-operator/blob/master/ImportTypes.dhall ?  It does seem like a nice paradigm. 


I was thinking, perhaps we can do this nested ala:

```
.
├── types
│   ├── io
│   │   ├── k8s
│   │   │   ├── api
│   │   │   │   ├── apps
│   │   │   │   │   ├── v1
│   │   │   │   │   │   ├── Deployment.dhall
│   │   │   │   │   │   └── DeploymentSpec.dhall
│   │   │   │   │   └── v1.dhall
│   │   │   │   ├── apps.dhall
│   │   │   │   ├── rbac
│   │   │   │   ├── rbac.dhall
│   │   │   │   ├── storage
│   │   │   │   └── storage.dhall
│   │   │   └── api.dhall
│   │   └── k8s.dhall
│   └── io.dhall
└── types.dhall
```
```dhall
# types.dhall
{ io = ./io.dhall }
```

```dhall
# k8s.dhall
{ api = ./api.dhall }
```

```dhall
# api.dhall
{ apps = ./apps.dhall
, rbac = ./rbac.dhall
, storage = ./storage.dhall
...
 }
```
```dhall
# apps.dhall
{ Deployment = ./Deployment.dhall
, DeploymentSpec = ./DeploymentSpec.dhall
}
```

You can then do something like:
```
let api = ./types.dhall
in  {... blah ... } : api.k8s.apps.v1.Deployment
```
Similar to how that's currently done in `dhall-prometheus-operator`

I'm not sure if the nested directory structure is neccssery per se.  Pro is that it allows people to just use
a small part of the API. We could also just opt for one big top-level `types.dhall`
that looks like:
```
io = { k8s = { apps = { v1 = {
   Deployment = ./io.k8s.apps.v1.Deployment.dhall
  , DeploymentSpec = ./io.k8s.apps.v1.DeploymentSpecc.dhall } } } } }
```

Not sure what is preferable. Thoughts?

Note that this problem is also present in  `typesUnion.dhall`.  Perhaps we should have one union per API namespace?

File: src/Main.hs
Comment (arianvp) [inline src/Main.hs]: This datatype (+ its json instances) is already in http://hackage.haskell.org/package/swagger2-2.3.1.1/
perhaps we can reuse it? That would solve the 'mixing of business logic and decoding' problem without having to have multiple modules ourselves

Comment (f-f) [reply src/Main.hs]: I didn't want to type this down so I went with `swagger2` at first, but the `swagger.json` we get from Kubernetes is broken and doesn't conform to the types there (the assumption here is that the implementation in `swagger2` is correct. The reason for the assumption is that we've been using it for years at work and it's the most correct Swagger library I've seen. On the other hand we've already seen cases in which k8s' swagger file is not the best)

File: stack.yaml
Comment (f-f) [reply stack.yaml]: We could keep both the stack file and the nix setup, as `dhall-haskell` does

File: package.yaml
Comment (f-f) [reply package.yaml]: Indeed, I just went with a random template, I planned to remove this. I'll add it to the TODO list

File: dhall-kubernetes.cabal
Comment (f-f) [reply dhall-kubernetes.cabal]: Yeah, this is part of the "remove `package.yaml`" task

**Comment by f-f** [comment]
@arianvp the approach I took here comes from the fact that I consider the `types.dhall` and `defaults.dhall` records part of "UX".
That is, they are not _strictly_ necessary (after all we managed without them just fine for a long time), but they help when doing things by reducing clutter: you could import all the objects by URL, but instead you import only one thing.
So introducing the full namespace would go in the other direction, as every time you have to reference an object you'd have to type the whole namespace

However, one might need objects that are not in these records (because they are on a different API version), so I propose we keep the current approach but also introduce a `typesNamespaced.dhall` and `defaultsNamespaced.dhall` (and a `typesUnionNamespaced.dhall`)

Some other notes:
- I wouldn't like a nested directory structure, as it's very hard to quickly navigate and it would be inconsistent anyways (see next point)
- splitting objects by namespace doesn't quite work, and the reason is that the objects naming is very inconsistent. For example, let's take the two objects `io.k8s.api.extensions.v1beta1.DeploymentCondition` and `io.k8s.api.apps.v1beta1.DeploymentCondition`. They are the same object, on the same API version, but in two different namespaces! Which one is the right one? Turns out the correct one is not the `io.k8s.api.extensions`, because that was the naming in development, and the "right" object (i.e. the one you'll find referenced by other objects) is the one in the `io.k8s.api.apps` namespace.
  There are tens of such weird cases - I spent several hours yesterday going through _all_ the objects before coming up with the current approach

File: src/Main.hs
Comment (f-f) [reply src/Main.hs]: +1 on this

File: src/Main.hs
Comment (f-f) [reply src/Main.hs]: Yeah that's in the TODO list

**Comment by arianvp** [comment]
I don't fully understand the second point.  But I'm fine with not nesting the directories so let's do as you suggest

**Comment by f-f** [comment]
@arianvp sorry I think I misunderstood your previous comment.
Updated take: let's keep the `types.dhall` as is, and have the nested records as you propose in `typesFull.dhall` (or `typesNamespaced.dhall`).
Performance might be bad compared to a flat record though. Also I'd like not to include this change in this PR, as this is already huge and it can be added later

File: package.yaml
Comment (f-f) [reply package.yaml]: Removed this and added version bounds!

**Comment by ari-becker** [comment]
> @arianvp the approach I took here comes from the fact that I consider the `types.dhall` and `defaults.dhall` records part of "UX".
> That is, they are not _strictly_ necessary (after all we managed without them just fine for a long time), but they help when doing things by reducing clutter: you could import all the objects by URL, but instead you import only one thing.
> So introducing the full namespace would go in the other direction, as every time you have to reference an object you'd have to type the whole namespace

Agree with this reasoning. Typing out the whole namespace for each record and type gets old fast. Would probably end in import style let-expressions (e.g. `let Deployment = (./path/to/Imports.dhall).Kubernetes.io.k8s.api.apps.v1.Deployment`) which feels wrong. In `dhall-prometheus-operator` I settled on that paradigm because, like you said @arianvp it's a nice paradigm, and it most closely matches the Kubernetes YAML objects (i.e. `apiVersion: apps/v1 kind: Deployment` translates to `Kubernetes.apps.v1.Deployment`) but it's also hand-coded so maybe it doesn't fit here, where you're trying to build this through a script and as such it's going to pick up everything in the OpenAPI document.

>     * splitting objects by namespace doesn't quite work, and the reason is that the objects naming is very inconsistent. For example, let's take the two objects `io.k8s.api.extensions.v1beta1.DeploymentCondition` and `io.k8s.api.apps.v1beta1.DeploymentCondition`. They are the same object, on the same API version, but in two different namespaces! Which one is the right one? Turns out the correct one is not the `io.k8s.api.extensions`, because that was the naming in development, and the "right" object (i.e. the one you'll find referenced by other objects) is the one in the `io.k8s.api.apps` namespace.

Question : because the users of this library are presumably Kubernetes end-users, does the `extensions.v1beta1` namespace have to be included? Certainly it would reduce clutter (and probably improve performance, since the records would be smaller) to only provide the "real" types and records?

**Comment by arianvp** [comment]
> Think about how we are going to build this

Lets keep everything in this Repo.  Dhall langauge is slowly starting to stabilize, so it doesn't matter if we 
still use an old dhall version for a while. Allows us to choose ourselves when we migrate to a new version of `dhall-haskell` and keeps stuff simple

File: src/Main.hs
Comment (f-f) [reply src/Main.hs]: I just realized that we already have an issue to track this: #8

**Comment by f-f** [comment]
With this latest changes I'm pretty sure now #46 is properly implemented and working.

This is how a simple deployment looks like with the new schemas:
```hs
let types =
      ../types.dhall sha256:1ac84a230c0fdd003b4db05157b11329d273217cc6837083eb30c9288a67c87f

let defaults =
      ../defaults.dhall sha256:2e3d2b5d5fb52d08b078e803fa88ac1ed0f788c6a4d6342621a76eb77afe40f4

let deployment
    : types.Deployment
    =   defaults.Deployment
      ⫽ { metadata =
            defaults.ObjectMeta ⫽ { name = "nginx" }
        , spec =
            Some
            (   defaults.DeploymentSpec
              ⫽ { replicas =
                    Some 2
                , template =
                      defaults.PodTemplateSpec
                    ⫽ { metadata =
                          defaults.ObjectMeta ⫽ { name = "nginx" }
                      , spec =
                          Some
                          (   defaults.PodSpec
                            ⫽ { containers =
                                  [   defaults.Container
                                    ⫽ { name =
                                          "nginx"
                                      , image =
                                          Some "nginx:1.15.3"
                                      , ports =
                                          [   defaults.ContainerPort
                                            ⫽ { containerPort = 80 }
                                          ]
                                      }
                                  ]
                              }
                          )
                      }
                }
            )
        }

in  deployment
```

And the new `omitEmpty` flag is so nice:
```bash
$ dhall-to-yaml --omitEmpty <<< "./examples/deploymentSimple.dhall"
apiVersion: apps/v1
kind: Deployment
spec:
  template:
    spec:
      containers:
      - image: nginx:1.15.3
        name: nginx
        ports:
        - containerPort: 80
    metadata:
      name: nginx
  replicas: 2
metadata:
  name: nginx
```

**Comment by f-f** [comment]
@arianvp any idea why Hydra shows a passing CI? The CI should fail for this branch, as all the files in `default` disappeared

**Comment by arianvp** [comment]
The CI _generates_ the default folder as part of its build step. The result of the `dhall-kubernetes` derivation is a `default` and `types` folder by calling `convert.py`.  The derivation actually ignores the `default` and `types` folder in the repo to ensure purity and always generates them from scratch

`generate.sh` builds the derivation and copies over the generated files to the root of the repo for convenience, if the derivation succeeds to build 

So before the tests are run, the `default` folder is repopulated hence the tests succeed :p. See the `Makefile`

We should modify the `Makefile` to not call `convert.py` but instead call our Haskell script.


Since we can't build our haskell script with Nix yet, we could hack out the types generation for now, so you can manually test stuff

In the makefile https://github.com/dhall-lang/dhall-kubernetes/blob/master/Makefile

change the `build` rule to:

```
build:
    echo lol nothing
```

and in the derivation https://github.com/dhall-lang/dhall-kubernetes/blob/master/nix/dhall-kubernetes.nix
make it so that the derivation doesn't ignore the `../types` and `../default` folder.

that should cause those folders even if generated by hand to end up in the nix derivation and be available for the `check-sources.sh` script in the `check` phase of the `Makefile` until
we fix packaging of the haskell file

**Comment by arianvp** [comment]
@f-f  The build uses haskell now! :confetti_ball:  The tests now fail; as expected

**Comment by arianvp** [comment]
@f-f I fixed the build :)

**Review by arianvp — APPROVED** [review]

**Comment by f-f** [comment]
@arianvp thanks a lot! 😍
I'll break CI again while trying to fix the examples 😄

File: dhall-kubernetes-generator.cabal
Comment (f-f) [inline dhall-kubernetes-generator.cabal]: @arianvp https://www.haskell.org/cabal/users-guide/developing-packages.html#accessing-data-files-from-package-code

Comment (f-f) [reply dhall-kubernetes-generator.cabal]: TL;DR: you want it to refer to static data without using Template Haskell (it includes other stuff too)

But you're right that we're not using it here 😄

**Comment by f-f** [comment]
@arianvp @ari-becker I rewrote the readme with the new examples, and documented the pattern for #54, could you proofread before we merge? 😊

@arianvp CI is failing because hash mismatches; I generated all the hashes locally with dhall `1.21.0`, and we should be using that one in the nix build, but it looks like we're not? 🤔

**Comment by f-f** [comment]
[First release done](https://github.com/dhall-lang/dhall-kubernetes/releases/tag/v1.0.0) 🎉, we are ready to merge this as soon as we get CI green 😄

**Comment by arianvp** [comment]
@f-f did you manually `freeze` the files?

We should adjust `Makefile` such that it freezes the dhall files...
because `types` and `defaults` should be generated by `./scripts/generate.sh` and that currently produces a set of unfreezed files :P

remember that the tests and the build run in a sandbox that exclude the `types` and `defaults` folder on purpose...  Such that `./scripts/generate.sh` always puts out a 'clean' build that doesn't succeed 'accidentally'

**Comment by f-f** [comment]
@arianvp the `types` and `defaults` I did freeze by hand in the last commit (and I'll push a fix to the haskell implementation so that it freezes the files when generating them), but the failures come from the examples - which I also froze manually, but they are supposed to be human curated so we have to.

**Comment by arianvp** [comment]
Looking at the diff on the examples:

```
apiVersion: apps/v1
kind: Deployment
spec:
  revisionHistoryLimit: 10
  selector:
    matchExpressions: []
    matchLabels:
      app: nginx
  strategy:
    rollingUpdate:
      maxSurge: 5
      maxUnavailable: 0
    type: RollingUpdate
  template:
    spec:
      initContainers: []
      tolerations: []
      containers:
      - image: nginx:1.15.3
        command: []
        args: []
        imagePullPolicy: Always
        env: []
        volumeMounts: []
        resources:
          limits:
            cpu: 500m
          requests:
            cpu: 10m
        name: nginx
        ports:
        - containerPort: 80
        envFrom: []
        volumeDevices: []
      readinessGates: []
      imagePullSecrets: []
      hostAliases: []
      volumes: []
    metadata:
      finalizers: []
      ownerReferences: []
      name: nginx
      labels:
        app: nginx
  replicas: 2
metadata:
  finalizers: []
  ownerReferences: []
  name: nginx
```

We forgot that not only we should omit `None`'s but also empty lists... is that something we can fix in `dhall-to-yaml` ?

**Comment by f-f** [comment]
@arianvp I did fix it some days ago 🙂 https://github.com/dhall-lang/dhall-haskell/pull/872

But now that I think of it CI will try to compare the committed yamls (which I generated with the patch above) with the newly generated ones - which might include empty lists - so it will likely fail. Should we pin `dhall-haskell` to the commit with the patch or should we wait for the next release?

**Comment by arianvp** [comment]
~~Let's just pin to that commit for now.   ~~


Edit: perhaps we should wait til the next dhall release. because people will probably install `dhall` from `hackage` and not from that pinned commit, and might be surprised that the output is different :P

File: README.md
Comment (ari-becker) [inline README.md]: I know this is necessary until it's merged in, but you should probably replace the commit hash with a tag that will exist, e.g. `v2.0.0`

**Review by ari-becker — COMMENTED** [review]
> @arianvp @ari-becker I rewrote the readme with the new examples, and documented the pattern for #54, could you proofread before we merge? 😊

Jeez, sorry I missed this somehow... LGTM though.

If part of the documentation is to try and show cool things that can be done with `dhall-kubernetes`, I'd just bring up a pattern I've found: sorry it's a little messy and hard to read:

```haskell
  λ(deployment : KubernetesTypes.apps.v1.Deployment)
→     Kubernetes.core.v1.Service
      { metadata = deployment.metadata }
    ⫽ { spec =
          Some
          (     Kubernetes.core.v1.ServiceSpec
              ⫽ { ports =
                    Some
                    ( Prelude.`Optional`.fold
                      KubernetesTypes.apps.v1.DeploymentSpec
                      deployment.spec
                      (List KubernetesTypes.core.v1.ServicePort)
                      (   λ ( spec
                            : KubernetesTypes.apps.v1.DeploymentSpec
                            )
                        → Prelude.`Optional`.fold
                          KubernetesTypes.core.v1.PodSpec
                          spec.template.spec
                          (List KubernetesTypes.core.v1.ServicePort)
                          (   λ ( spec
                                : KubernetesTypes.core.v1.PodSpec
                                )
                            → Prelude.`List`.concatMap
                              KubernetesTypes.core.v1.Container
                              KubernetesTypes.core.v1.ServicePort
                              (   λ ( container
                                    : KubernetesTypes.core.v1.Container
                                    )
                                → Prelude.`Optional`.fold
                                  (List KubernetesTypes.core.v1.ContainerPort)
                                  container.ports
                                  (List KubernetesTypes.core.v1.ServicePort)
                                  (   λ ( portList
                                        : List
                                          KubernetesTypes.core.v1.ContainerPort
                                        )
                                    → Prelude.`List`.map
                                      KubernetesTypes.core.v1.ContainerPort
                                      KubernetesTypes.core.v1.ServicePort
                                      (   λ ( containerPort
                                            : KubernetesTypes.core.v1.ContainerPort
                                            )
                                        →     Kubernetes.core.v1.ServicePort
                                              { port =
                                                  containerPort.containerPort
                                              }
                                            ⫽ { name =
                                                  containerPort.name
                                              , protocol =
                                                  containerPort.protocol
                                              , targetPort =
                                                  let IntOrString =
                                                        < Int :
                                                            Natural
                                                        | String :
                                                            Text
                                                        >

                                                  in        if Prelude.`Optional`.null
                                                               Text
                                                               containerPort.name

                                                      then  Some
                                                            ( IntOrString.Int
                                                              containerPort.containerPort
                                                            )

                                                      else  Prelude.`Optional`.map
                                                            Text
                                                            IntOrString
                                                            (   λ ( name
                                                                  : Text
                                                                  )
                                                              → IntOrString.String
                                                                name
                                                            )
                                                            containerPort.name
                                              }
                                          : KubernetesTypes.core.v1.ServicePort
                                      )
                                      portList
                                  )
                                  ( [] : List
                                         KubernetesTypes.core.v1.ServicePort
                                  )
                              )
                              spec.containers
                          )
                          ([] : List KubernetesTypes.core.v1.ServicePort)
                      )
                      ([] : List KubernetesTypes.core.v1.ServicePort)
                    )
                , selector =
                    deployment.metadata.labels
                , type =
                    Some "ClusterIP"
                }
            : KubernetesTypes.core.v1.ServiceSpec
          )
      }
  : KubernetesTypes.core.v1.Service
```

And what that does is basically generate a `Service` from a `Deployment`, mapping each of the ports of the `Deployment` to be exposed through a `Service`. If ports or added or changed in the `Deployment`, then those changes will be reflected in the `Service`.

It'd have to be updated for API v2 of course, and maybe it should be in some kind of function library for `dhall-kubernetes`, but yeah, I thought it was pretty cool when I finally got it to work.

**Comment by arianvp** [comment]
LGTM!

**Comment by f-f** [comment]
@ari-becker looks great! I'd say more docs is better, so let's include useful patterns too. @arianvp was planning a documentation site after we merge this in, so this would be definitely welcome there

(small sidenote: we now provide the `IntOrString` type too in the `types.dhall` record, so no need to redeclare it inline)

**Pull request merged**
